python/dgl/heterograph.py

```diff
@@ -0,0 +1,1214 @@
+"""Classes for heterogeneous graphs."""
+
+class DGLBaseHeteroGraph(object):
+    """Base Heterogeneous graph class.
+
+    A Heterogeneous graph is defined as a graph with node types and edge
+    types.
+
+    If two edges share the same edge type, then their source nodes, as well
+    as their destination nodes, also have the same type (the source node
+    types don't have to be the same as the destination node types).
+
+    Parameters
+    ----------
+    metagraph : NetworkX MultiGraph or compatible data structure
+        The set of node types and edge types, as well as the
+        source/destination node type of each edge type is specified in the
+        metagraph.
+        The edge types are specified as edge keys on the NetworkX MultiGraph.
+        The node types and edge types must be strings.
+    node_types : list
+        The node types.
+    edge_connections_by_type : dict
+        Specifies how edges would connect nodes of the source type to nodes of
+        the destination type in the following form:
+
+            {edge_type: (source_node_id_tensor, destination_node_id_tensor)}
+
+        where
+
+        * ``source_node_id_tensor`` and ``destination_node_id_tensor`` are
+          IDs within the source and destination node type respectively.
+        * ``edge_type`` is a triplet of
+
+              (source_node_type_name,
+               destination_node_type_name,
+               edge_type_name)
+
+    Examples
+    --------
+    Suppose that we want to construct the following heterogeneous graph:
+
+    .. graphviz::
+
+       digraph G {
+           Alice -> Bob [label=follows]
+           Bob -> Carol [label=follows]
+           Alice -> Tetris [label=plays]
+           Bob -> Tetris [label=plays]
+           Bob -> Minecraft [label=plays]
+           Carol -> Minecraft [label=plays]
+           Nintendo -> Tetris [label=develops]
+           Mojang -> Minecraft [label=develops]
+           {rank=source; Alice; Bob; Carol}
+           {rank=sink; Nintendo; Mojang}
+       }
+
+    One can analyze the graph and figure out the metagraph as follows:
+
+    .. graphviz::
+
+       digraph G {
+           User -> User [label=follows]
+           User -> Game [label=plays]
+           Developer -> Game [label=develops]
+       }
+
+    Suppose that one maps the users, games and developers to the following
+    IDs:
+
+        User name   Alice   Bob     Carol
+        User ID     0       1       2
+
+        Game name   Tetris  Minecraft
+        Game ID     0       1
+
+        Developer name  Nintendo    Mojang
+        Developer ID    0           1
+
+    One can construct the graph as follows:
+
+    >>> import networkx as nx
+    >>> metagraph = nx.MultiGraph([
+    ...     ('user', 'user', 'follows'),
+    ...     ('user', 'game', 'plays'),
+    ...     ('developer', 'game', 'develops')])
```

Comment:
Is this networkx syntax? the last element in the tuple represents edge type?

Reply:
Yes